lib/token-store.js

```diff
@@ -8,196 +8,1110 @@
+    }
+
+    return createCursor(tokens, comments, imap, start, end, includeComments, filter, 0, count);
+}
+
+/**
+ * Creates the cursor iterates tokens with options.
+ *
+ * @param {Token[]} tokens - The array of tokens.
+ * @param {Comment[]} comments - The array of comments.
+ * @param {Object} imap - The map from locations to indices in `this.tokens`.
+ * @param {number} start - The start location of the iteration range.
+ * @param {number} end - The end location of the iteration range.
+ * @param {Function|Object} opts - The option object. If this is a function then it's `opts.filter`.
+ * @param {boolean} [opts.includeComments] - The flag to iterate comments as well.
+ * @param {Function|null} [opts.filter=null] - The predicate function to choose tokens.
+ * @param {number} [opts.count=0] - The maximum count of tokens the cursor iterates. Zero is no iteration for backward compatibility.
+ * @returns {Cursor} The created cursor.
+ * @private
+ */
+/**
+ * Creates the cursor iterates tokens with options.
+ *
+ * @param {Token[]} tokens - The array of tokens.
+ * @param {Comment[]} comments - The array of comments.
+ * @param {Object} imap - The map from locations to indices in `this.tokens`.
+ * @param {number} start - The start location of the iteration range.
+ * @param {number} end - The end location of the iteration range.
+ * @param {number} [beforeCount=0] - The number of tokens before the node to retrieve.
+ * @param {boolean} [afterCount=0] - The number of tokens after the node to retrieve.
+ * @returns {Cursor} The created cursor.
+ * @private
+ */
+function createCursorWithPadding(tokens, comments, imap, start, end, beforeCount, afterCount) {
+    if (typeof beforeCount === "undefined" && typeof afterCount === "undefined") {
+        return new TokenCursor(tokens, comments, imap, start, end);
+    }
+    if (typeof beforeCount === "number" || typeof beforeCount === "undefined") {
+        return new PaddedTokenCursor(tokens, comments, imap, start, end, beforeCount, afterCount);
+    }
+    return createCursorWithCount(tokens, comments, imap, start, end, beforeCount);
+}
+
+class TokenStore {
+
+    /**
+     * Initializes this token store.
+     *
+     * ※ `comments` needs cloning for backward compatibility.
+     * After this initialization, ESLint removes a shebang's comment from `comments`.
+     * However, so far we had been concatenating 'tokens' and 'comments',
+     * so the shebang's comment had remained in the concatenated array.
+     * As the result, both `getTokenOrCommentAfter` and `getTokenOrCommentBefore`
+     * methods had been returning the shebang's comment.
+     * And some rules depends on this behavior.
+     *
+     * @param {Token[]} tokens - The array of tokens.
+     * @param {Comment[]} comments - The array of comments.
+     */
+    constructor(tokens, comments) {
+        this.tokens = tokens;
+        this.comments = comments.slice(0); // Needs cloning for backward compatibility!
```

Comment:
I noticed shebang's comment (`///usr/bin/env node`) remains in token store currently. I think this is a bug, but I keep it to avoid breaking changes.

> ※ `comments` needs cloning for backward compatibility.
> After this initialization, ESLint removes a shebang's comment from `comments`.
> However, so far we had been concatenating 'tokens' and 'comments',
> so the shebang's comment had remained in the concatenated array.
> As the result, both `getTokenOrCommentAfter` and `getTokenOrCommentBefore`
> methods had been returning the shebang's comment.
> And some rules depend on this behavior.